oktapam/resource_ad_user_sync_task_settings.go

```diff
@@ -0,0 +1,231 @@
+	"github.com/okta/terraform-provider-oktapam/oktapam/constants/descriptions"
+	"github.com/okta/terraform-provider-oktapam/oktapam/logging"
+	"github.com/okta/terraform-provider-oktapam/oktapam/utils"
+)
+
+func resourceADUserSyncTaskSettings() *schema.Resource {
+	/** ADUserSyncTaskSettings is immutable. Only updatable attributes are active status and task frequency/start time.
+		ADUserSyncTaskSettings resource has update endpoint, but internally it's a transactional delete & then create
+	    operation. Terraform resource id change and deletion of existing resource confuse terraform. Not sure if there is
+	    any way to handle it properly. For now, I have marked all the immutable attributes as "ForceNew". If update
+	    ADUserSyncTaskSettings endpoint has additional logic then we need to revisit this.
+	    TF Reference - https://www.terraform.io/plugin/sdkv2/schemas/schema-behaviors#forcenew
+	*/
+	return &schema.Resource{
+		Description:   descriptions.ResourceADUserSyncTaskSettings,
+		CreateContext: resourceADUserSyncTaskSettingsCreate,
+		ReadContext:   resourceADUserSyncTaskSettingsRead,
+		UpdateContext: resourceADUserSyncTaskSettingsUpdate,
+		DeleteContext: resourceADUserSyncTaskSettingsDelete,
+		Schema: map[string]*schema.Schema{
+			attributes.ADConnectionID: {
+				Type:        schema.TypeString,
+				Required:    true,
+				Description: descriptions.ADConnectionID,
+			},
+			attributes.Name: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: descriptions.Name,
+			},
+			attributes.ID: {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			attributes.Frequency: {
+				Type:         schema.TypeInt,
+				Required:     true,
+				ValidateFunc: validation.IntBetween(1, 24),
+				Description:  descriptions.ADUserSyncTaskSettingsFrequency,
+			},
+			attributes.StartHourUTC: {
+				Type:         schema.TypeInt,
+				Optional:     true,
+				ValidateFunc: validation.IntBetween(0, 23),
+				Description:  descriptions.ADUserSyncTaskSettingsStartHourUTC,
+			},
+			attributes.BaseDN: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: descriptions.ADUserSyncTaskSettingsBaseDN,
+			},
+			attributes.LDAPQueryFilter: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: descriptions.ADUserSyncTaskSettingsLDAPQueryFilter,
+			},
+			attributes.UPNField: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: descriptions.ADUserSyncTaskSettingsUPNField,
+			},
+			attributes.SIDField: {
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: descriptions.ADUserSyncTaskSettingsSIDField,
+			},
+			attributes.IsActive: {
+				Type:        schema.TypeBool,
+				Required:    true,
+				Description: descriptions.ADUserSyncTaskSettingsIsActive,
+			},
+			attributes.RunTest: {
```

Comment:
I am going to check more on this. I think the ideal way to handle it is to suppress diff for this attribute. This attribute is only required while creating resource and then it's never returned by the api while reading it back. Computed is used for attributes which are not known at creation time and returned back after creating the resource for example resource identifiers.

Reply:
Can't find a primitive that does "ignore this attribute for update trigger calculation, but use it as it is specified for create". There is DiffSuppressFunc, which can avoid triggering update, but will use the false value always when we recreate. Note that at present the only problem is the noise during plan when the value is true, ideally it should not be there, and the execution should probably not enter provider logic, but at least no platform API call is made in that case.